### ISSUE TITLE
Confirmation or rejection of implementation of analogic urxvt features

### ISSUE THREAD

**Issue by etosan** [alacritty/alacritty#3749]
### Introduction

Sorry for barging in but when I saw #3711, I had to chime in. 

This has been sitting in my drawer since I started experimenting with `alacritty` quite a few months ago. And this also will be long, so beware and brace for longer read please.

It is essentially an comprehensive essay what on would be needed to make `alacritty` capable of really replacing `rxvt-unicode` for it's former power users. `urxvt` is deeply entrenched into systems of certain users and rightly so.

I have been using `alacritty` some months but it's still too weak to actually replace my trusty workhorse of `rxvt-unicode`. 

But as things are now I would be willing to do it. There is nothing wrong with `urxvt` (besides **perl** and few other things) but `alacritty`'s renderer really is something else.

I know I am not alone. We are few and far between, but we exist, and we treat our terminals as gospels. Let's be frank, despite future being "now" state of most of the terminal emulators on the market is utter shit (there is simply no other way to call it).

Somebody (@Deewiant) is messing with dirty insides and iron has to be forged while hot and I believe momentum behind #3711 (and few others) could be actually (ab)used to get this ball rolling. So I believe time is right to post this.

Once in a while I give `alacritty` occasional testride, but I have to revert back to `urxvt` often. While keyboard handling and other areas still have some very minor glitches, overall, `alacritty` is one of the best modern(!) contenders (and probably only one at that) capable to position itself on "actually useful terminal (for power users)" spot. All that despite it's youth.

I guess that in current state `alacritty` can now can be considered very successful research & development prototype, that has succeeded at reaching primary research goals (mainly fluid hw. accelerated glyph rendering and scrolling). 

**Now time has come for it to pick next level direction and for it's devs to actually take a stance: "will these features be eventually (and properly) implemented or never"?**

I am asking this, so that those "few of us" that are stuck with `urxvt` know whether we should be wasting our time (and hope) with this "little" and very interesting toy or should we return back into our caves, to our old and trusted tool.

Keep in mind this is written from point of view supervision (daemontool,runit,s6) AND `xorg` user. I guess "daemon mode" discussed, does not translate well to waylands, but what do I know. And yeah despite sounding like super rare user animal - I believe issues I raise here end up affecting whole xorg (at least) userbase. I would like to know where we will end up at.

Picking up challenges about which I write here, and solving them properly, I believe, also gives chance to rustaceans responsible to prove the modularity and greatnes of their language of choice ;). 

If filthy c++ based `rxvt-unicode` can do it, there must be way for `alacritty` to do it better in rust, too. Terminal is a package that many of us interact daily and it shapes our perception of systems as "a whole" extremely.

Also note that despite sounding (probably) confrontational, I don't want to diss `alacritty` at all (I am actually a very big fan), and I would like it to succeed as power user's terminal. 

Sections following:
 - proper daemon mode
 - proper multiprocess model with no doubleforks
 - scripting layer and APIs

---

###  Proper daemon mode

At it's current state `alacritty` is quite fine as single tmux window host, or singular terminal on "cutesy" unix-porn-y desktop. Users stuck in such use cases can bask in fotons produced by it's hw. accelerated rendering and can also enjoy it's raw speed. 

Using it this way I got closest to raw DOS VGA rendering + PS/2 input feel in years. `rxvt-unicode` used to be very good in this regard and its rendering and responsiveness is still very fine, but I am noticing (and I am not alone) that it feels somehow different last few years. This is the main reason why my eyes are set on `alacritty`.

However once you move past "single console" ranger/vim usage to the actual 8h-12h workday level load, `alacritty` starts becoming, sadly, rather annoying. And quickly.

If you are managing 100s of pet servers, many in clusters, interconnected in complicated sdn/hw/vlan mixed topologies, in various state of "disarray", often behind several jump boxes and perimeters ("because of business reasons"), cute little windows with decorations and mouse clicks won't cut it anymore. You will switch to tiling window manager, and are smashing keyboard like mad to keep stuff rolling and working. 

Regarding window management, `alacritty`, despite it being opengl program, holds surprisingly well. For the love of god don't add tabs and other GUI bullshit that can be left to proper tools, designed for that.

Unfortunately, in the work environment I described above, terminal window is very cheap and throwaway resource and because of that new terminal window-spin up time is very important metric. 

Practically (despite what you might have heard from various vocal "users") there has been one and only one, singular winner, in this niche area, for decades, and that is `rxvt-unicode`'s daemon mode. 

With `rxvt-unicode` in this mode, instead of having one process per terminal window model, you get terminal "window drawing daemon" model, where all terminal windows are spawned, drawn and managed by single `urxvtd` process. 

Also all `pty` `master/slave` pairs of all the "shell" children belong to this "daemon terminal" process as well. 

This "daemon" then handles both drawing and data pumping on behalf of all it's "CLI clients". Finally, you get `urxvtc`, a "client" tool which is basically glorified message sender, that packs `cwd`, `env` and `argv` into new "spawn term" message and sends it to actual "daemon" for spawning. 

New "urxvt terminals" spawned this way pop-in instantly. Naturally, because they are just new windows of a single, big and already existing process. Spawning a shell linked to `pts` end of the window is thus much faster, than spawning the **new terminal process and shell process pair**.

Despite what various "siloed (pretend) systems" people are telling you and despite `fork()/exec()` being really several orders of magnitude faster and cheaper than `CreateProcess()` on Windows, what happens after exec() is not that fast on linux (or any modern unix for that matter) anymore. We live in .so/.dll era and relinking new `alacritty` each time you want a new window takes considerable CPU and time. On this box query `lddtree.sh -l =alacritty | wc -l` gives me 36 non-trivial solibs to link each time I spawn a window. Resulting process then needs to load and parse config, fonts and other crap (I imagine).

But there are more inefficiencies. Each new `alacritty` process gets new static data tables, stuff that cannot be shared and what's worst: quite a few of threads (which is sad and clutters detailed process listings - it's quite nontrivial to write commands to see CLI child process threads but not `alacritty` own ones - but for the sake of fast response and beautiful rendering I am willing to live with this annoyance). 

I bet even OpenGL subsystem would be more grateful if single " alacritty daemon" could share multiple terminal window buffers and OpenGL contexts within single process (font buffers and other OpenGL stuff too). I also know that multi window OpenGL applications are a thing, so this must be, certainly, doable.

Finally I am also pretty sure some magic algorithmic things could be applied in daemon mode to optimize `pty` data pumps too.

Unfortunately using `alacritty` like if it was `urxvt`currently, you'll end up with dozens of processes, and insane number of threads all over the place. Lets say you want to observe subprocess tree of specific shell, searching just for this process in a tree becomes annoying.

`alacritty` supporting this mode would also invalidate several other issues, for example #2735, as then there is only one client with "lotsof x windows" instead of "lotsof x clients with one x window each". Not to mention other resources that could be conserved considerably and process tree that would get visibly de-cluttered during busy workday.

Some people will tell you that with this design, terminal daemon crashing is tragedy waiting to happen.  Those of us that use urxvtd daemon daily and probably spawn and tear down more terminal windows in a busy workday than your casual user spawns in a week or a month, can assure you, this design is valid and can lead to rock solid experience. I am using `urxvtd` heavily for more than 13 years daily and I have had like 2-3 unexpected crashes during that entire time. I have seen gnome terminal crash more often in my life, than `urxvtd`. Bonus point: isn't rust actually a language that has safety built in and that is designed to allow programmers to write crash-free software exactly like this?

Proper daemon mode is becoming probably most important feature `alacritty` will need to have. Especially in it's conquest for speed.

To recapitulate this mode, to be at least considered seriously, when compared against already available alternatives, must be able to:
  - "lead" unlimited number of "shell" children, eg. manage unlimited number of `pts` bound child processes (I believe this should be non issue with current design)
  - host unlimited number of terminal windows (think like this several 100s of windows spanning several virtual desktops)
  - provide proper ipc server socket (for the love of god and unborn children NOT dbus, just plain s-------- file)
  - provide ability to specify different than default ipc server socket (so that you can have multiple "daemon" instances, practice some power users I know use to partition their shells)
  - have easily usable 'spawn-new-term' "client" (so that it can be scripted from session login scripts, xinitrc and whatnot) that can connect to any socket specified
  - provide proper foreground daemon mode (so that daemon can be supervised, by systemd (cool kids), runit, s6 or what have you) 

Although this just seems like `urxvt` feature parity, this is actually minimal required feature set needed for proper **and usable** `urxvt daemon`-like implementation. With this built into `alacritty`, we can finally use it realistically in high terminal count situations. 

Regarding advancing daemon mode further than `rxvtd-unicode` (eg. research), thanks to `nosh` package we now know, that splitting deamon into two separate parts (perhaps two processes `renderer` and `pts holder`) could be used to increase resiliency of given setup. Should `renderer` crash, if `pts holder` has the means to restart it (fork()?), and a way to communicate pty fds and linkages to it, no "CLI client" will be lost. 

Whether such separation is worth the hassle is not my call to make. This brings up quite a lot of original questions and problems to solve though (what should happen to 100s of newly respawned windows suddenly arriving into xorg session, windowing and virtual desktops wise). But such architecture can also be used as argument to silence critics regarding perceived "dangerousness" of daemon mode and break new grounds for `alacritty` as research device.

Anyway this "daemon" mode is becoming most essential feature missed by `urxvt` expats by the day.

---

### Proper multiprocess model with no doubleforks

Whether devs will go or not with daemon mode, second major pain point is children handling. I believe there are some inconsitencies currently.

As seen in #3711, there are multiple child spawning and child reaping (synchronisation) points within `alacritty` codebase, with different semantics. More over for some inexplicable reason `alacritty` can even doublefork(!). 

Normal user level software should never doublefork for whatever reason (unless it is specialest of the special reasons and cases, and to me only `tmux` qualifies as an only example, and only in very, very few situations). 

Doubleforking itself is an old hack that has no place in any software written post year 2000.

There seems to be some confusion in `alacritty` how to do it properly. Whether it is daemon mode or single window terminal mode, `alacritty` should always, for any reason, spawn it's children directly (no matter how short lived) and should never doublefork at any circumstamce. Finally, it should, at all times, behave as fully qualified reaper (eg init). 

As author of `s6` suite puts it in his `s6-svscan` page `alacritty` should reap all zombie children it has in following way (replace `s6-svscan` with `alacritty` there):
> The reaper acknowledges (via some wait() function), **without blocking**, **every terminated child** of `s6-svscan`, **even ones it does not know it has**. This is especially important when `s6-svscan` is run as process 1. 

This reaping mechanics will ensure that `alacritty`, **knows** when it's main child (the one that it basically supervises, and first one that gets connected to it's `pts` (usually shell)), or any other short lived child for that matter, terminates, but also **allows it to handle children it got by some other (often unexpected) means**. This is especially crucial in daemon mode, but is indeed proper unix behaviour for all master/slave multiprocessing (daemon or not) constructions. 

For example, I, in certain service instances, often explicitly mark `utxvtd`, `sshd` or `tmux` process as "subreaper". After marking process by this special Linux/FreeBSD API, kernel will reparent any orphaned processes of a given "subreaper" grandchildren right under it, eg. "under" the given `urxvtd`, `sshd` or `tmux` instance.

This is useful trick to track origins of processes in the system and good way to use terminal and daemon processes as firewall/perimeter points in process tree.  

While this usecase of mine is rather specific hack, it illustrates reason to have unified child reaping handling. There are other ways to obtain reapable children we don't know about (and new might appear in the future due containers and whatever), and completely unified, predictable and robust event driven handling of all children zombies at all times is required in alcritty.

It doesn't need to be stressed that thanks to proper and unified child handler design, all "daemons" mentioned in my "subreaper" example handle case of being init-like reapers just fine. Should any reparented grandchild `exit()` sooner than our daemon process, any properly written daemon will reap it without leaving a zombie lingering and without crash or confusion. This behaviour is so crucial, that even shells do it properly. As such `alacritty` should ensure it's true for it as well.

---

### Scripting layer and APIs

This one is the one that is the most "out there". Although this might seem ludicrous to some, it isn't. It is essential feature for any advanced terminal.

Unfortunately `urxvt-unicode` is gold standard here again. 

What I believe would be needed most, is ability to create custom actions and have some means to hook them to keybinds and mousebinds, preferably completely arbitrarily.

Binding mechanisms must be free-form enough to allow binding of any conceivable key combination with any xkeyb modifiers combination pressed to be bound to any scripted action. In `urxvt`, unfortunately, exist some very few and specific sequences that are effectively unbindable, at least under certain conditions. Should `alacritty` decide to ever implement scripting layer, this must not happen, at all costs. 

Also certain bind expressions in `urxvt` are quite arcane and/or confusing (`S` == `Shift + KEY_S`), but after dealing with current binding system in `alacritty`, sequences used here seem unambigous.

Once we can script actions, next APIs we need most is direct `alacritty` graphical buffer control. This sounds very scary but it is not. I believe major scripted use cases for most `urxvt` power users are various pans and scrolls.

I personally am bound mostly to specialty scrolls (one line, 3/n lines, half page, page, scrollback start, scrollback middle, scrollback end, scrollback "line-bookmarks"), scrollback zap (drop and clear everything so far and reset term - several order of magnitude faster than calling reset in "target" and works across ssh sessions to managed switches and disk arrays too), cursor toggles, and character cells drawing control. Many of these subsystems are already in place but are not (yet) exposed to user. `urxvt-perls` package does some other stuffs too, but I think perl scripts used in it sould be first examined carefully to fish out only most important API control points.

All APIs mentioned so far are concerned with quickly getting around the walls of text and moving terminal viewport to important places. In my eyes these are first class of APIs in `urxvt`.

Second class of `urxvt` APIs are buffer searches (mostly) and manipulations. We use these to walk the `urxvt` buffers and fish out important stuff from them. To kickstart your imagination, thigs we fish out the most are often urls, paths, timestamps and crap like that. And of course we do freeform search strings and regexps a lot - but this can be deferred to external tools as well. 

I know `alacritty` already provides url detector, and I admit I haven't played with it much, so I don't know how generic it is, but as far as I know, it's unactivable from keyboard. This essentialy nullifies any potential power use. My own `url` selector fishes for `smb://`,`file://` based on various toggles, for example. With unscripted "data fishers" unacessible form keyboard, it is pretty hard to do thigs like that in `alacritty`.

Regarding the second class APIs, iterative search "walkers" and full scrollback buffer access and control is of utmost importance to us and builtin clipboard and cutbuffers access would be nice (in `urxvt` we have to rely on sub-processes, which is quite shitty honestly). Observe that `alacritty` already has many parts of the infrastructural code for this in place.

One of other more important features is dumping whole scrollback buffer to a file (or some other datastore - like externaly obtained pipe or string within internal perl runtime) for further analysis.

With these extensions in place I believe `alacritty` could confidently replace `rxvt-unicode` in 90% use cases more advanced users use it for.

I think scripting language should be the devs choice, but I would prefer if was not javascript. 

I also dislike perl, but one cannot get around that one in `urxvt`. Despite it's incredible string processing capabilities, it's simply too complex language for this task. 

I think lua(jit) which has pretty good string matching and string processing capabilities could perhaps be considered. It is pretty small.

Anway any language is fine as long as we can script our workflows.

---

### Conlusion

Reading this back after myself I realized how much stuff `alacritty` (or any other term for that matter) is missing to even come close to flexibility of `rxvt-unicode` - maybe this whole replacement idea is just pipedream?

**Comment by chrisduerr** [comment]
> New "urxvt terminals" spawned this way pop-in instantly. Naturally, because they are just new windows of a single, big and already existing process. Spawning a shell linked to pts end of the window is thus much faster, than spawning the new terminal process and shell process pair.

This is used a lot by people describing URxvt's daemon mode, but I've personally never made that experience. I don't see how a daemon mode would significantly improve startup time. Even if it was true in URxvt, it's very likely not the case for Alacritty. Almost the entire time spent in startup is the creation of the new window, it's not spawning a new process that takes this long. So all you'd be doing is just moving the spawning to a different place that you cannot `time`, without actually improving the initialization time.

It's also a much, much worse worst case scenario for people that do not use tmux. Now instead of losing one window when something bad happens, you manage to lose every single instance since your host Alacritty process crashes. That just seems like a very poor idea from a process isolation perspective.

That said, support for this is likely necessary to make Alacritty work nicely with strangely designed operating systems like macOS, so it's probably going to land sooner or later. But I don't expect it to improve anything significantly on Linux, expect for maybe some placebo.

> This reaping mechanics will ensure that alacritty, knows when it's main child (the one that it basically supervises, and first one that gets connected to it's pts (usually shell)), or any other short lived child for that matter, terminates, but also allows it to handle children it got by some other (often unexpected) means.

This makes the invalid assumption that Alacritty is interested in handling its children, which it is not. While the primary children like the shell are spawned and supervised by Alacritty, there is no interest in keeping any kind of connection to processes forked out to execute some arbitrary commands unrelated to Alacritty. We do not want to know anything about them and we do not want to be affected by anything they do.

> This is useful trick to track origins of processes in the system and good way to use terminal and daemon processes as firewall/perimeter points in process tree.

That's not what terminal emulators are though. If you want to do that, you should do it with a program designed to do it. Not some arbitrary program where you decide to put these processes. This does not illustrate a reason to have unified child reaping, it illustrates why we should do our best to stay as far away from this as possible. we do **not** want to get involved with this.

> Scripting layer and APIs

Alacritty should **never** support this in any way even close to URxvt. I've used URxvt and their plugin system is an absolute mess full of subpar and terrible software that people should stay away from as far as possible.

Many of the things handled by URxvt's common plugins are either part of Alacritty already, or currently worked on. Including features like keyboard URL handling, scrollback search and passing the scrollback to a file or buffer.

The idea of having plugins is really just a way to write low quality software that doesn't work well enough to be upstreamed in Alacritty. And just like URxvt is judged by the performance it has with common plugins enabled the same would be true for Alacritty. Alacritty's performance goals cannot be upheld with arbitrary code that is too poor to land upstream just running around doing random things it shouldn't be doing. If the software is good and the implementation is proper, it can be added to Alacritty, so there's no need for plugins.

So ultimately I don't really see much value in any of these URxvt "features" and landing them in Alacritty is unlikely. There's a bug for daemon mode (#241), which will likely be implemented because it's relatively straight-forward with multi window support, but not because it provides significant advantages. The other things are really more drawbacks than benefits.

**Comment by etosan** [comment]
Thank you @chrisduerr. This was very much needed. I still have few things to add and some questions.

> This is used a lot by people describing URxvt's daemon mode, but I've personally never made that experience. 

As in, "when you actually tried, you never observed the speed-up when using `urxvt` in this mode?", or as in "I don't see how this could possibly improve speed based on what I know about things?" ? 

I am just curious honestly. I started using `urxvt` in daemon mode on shitty 32-bit Turion system because normal `urxvt` was too slow to spawn there. Even today, when I have 8 core notebook with 32 GBs of RAM and 2 SSD-s, I see noticeable difference in new window spawning latency when using `urxvtc` instead of `urxvt` (when using window manager keybinds). Naturally, I do realize this does not mean that there is 100% certainty that  `alacritty` would benefit from this design as well. As you say, rust, OpenGL and other bindings between `alacritty` components might be too different and have much different hotspots, than `urxvt`, XLib and  C/C++ have, and might end up having completely different behavior and performance characteristics. Also `urxvt` is only terminal on market actually doing this (daemon mode), so it is essentially impossible to do any meaningful performance comparison.

> I don't see how a daemon mode would significantly improve startup time. Even if it was true in URxvt, it's very likely not the case for Alacritty. Almost the entire time spent in startup is the creation of the new window, it's not spawning a new process that takes this long. So all you'd be doing is just moving the spawning to a different place that you cannot `time`, without actually improving the initialization time.

So you are saying, that you believe, that worker forking strategy known for decades, well understood and used by nginx, php-fpm (not that it is quality software) and other daemons with similar setup (including urxvtd) is "certainly" not gonna do difference in `alacritty`'s case because graphics subsystem initialization cannot be sped up anyway, and other gains are minuscule compared to it?

Even if answer is yes, fair point. We are, after all, in opinion territory here, and nobody really knows for sure, unless it is tried (I now realize this will be original research in `alcritty`'s case). I dare to cast my vote in favor of daemonized mode being faster to spawn new terms, but you are right, only once feature lands we will know for sure.

> It's also a much, much worse worst case scenario for people that do not use tmux. Now instead of losing one window when something bad happens, you manage to lose every single instance since your host Alacritty process crashes. That just seems like a very poor idea from a process isolation perspective.

Well, if you want to go super-reliable route, as long as pts fds and masters will remain held somehow, "clients" won't even notice the crash I believe, but I understand the reasoning. I admit I might be mistaken myself, I have never written terminal emulator, I am only playing with building blocks.

`<sarcasm>`But I am surprised still, that rustaceans are not up to challenge of attempting to write "almost uncrashable software" and are not believing in reliability of their own tooling, while C/C++ persons, like `urxvt` author, are pretty confident in their ability to write code that can keep their daemon afloat. I was sure you would gulp the hook.`</sarcasm>` 

Joking aside, I understand general unwillingness to accept speed vs reliability trade off here. `urxvt` daemon mode is still advanced user feature and comes with big fat warning of this worst case you mention, so user is very well aware what the risk involved is.

> That said, support for this is likely necessary to make Alacritty work nicely with strangely designed operating systems like macOS, so it's probably going to land sooner or later. But I don't expect it to improve anything significantly on Linux, expect for maybe some placebo.

Understood. Let's see what happens in the future.

> This makes the invalid assumption that Alacritty is interested in handling its children, which it is not. While the primary children like the shell are spawned and supervised by Alacritty, there is no interest in keeping any kind of connection to processes forked out to execute some arbitrary commands unrelated to Alacritty. We do not want to know anything about them and we do not want to be affected by anything they do.

Yes this makes the invalid assumption that `alacritty` is interested in somehow "handling" (besides reaping that is) its children. This is not question of interest in children, but of decades known, "proper" unix program design. You don't need know anything about children, you just need to call proper wait() function variant in proper way to reap all children and not just single PID (shell) focused one. That's all that is needed and maybe you are already doing it. I just wanted to point out and stress that this behavior should be kept in place at all times. I don't speak rust and I don't know where to look, to verify what is being done.

> This does not illustrate a reason to have unified child reaping, it illustrates why we should do our best to stay as far away from this as possible. we do not want to get involved with this.

In retrospect I admit, maybe it was wrong example to pick and I rubbed you the wrong way, but what do you propose in case, for whatever whatever reason, zombie accumulates under `alacritty`? You declare that is "not your problem" and expect knowledgeable user to restart the terminal (so that said zombie can get "teleported" under `init` that will reap it without hassle properly) to zap it? Just so that we are clear, `yes` is valid response here.

> Alacritty should never support this in any way even close to URxvt. I've used URxvt and their plugin system is an absolute mess full of subpar and terrible software that people should stay away from as far as possible.
>
> The idea of having plugins is really just a way to write low quality software that doesn't work well enough to be upstreamed in Alacritty. And just like URxvt is judged by the performance it has with common plugins enabled the same would be true for Alacritty. 

My performance profile with "plugins" certainly does not match the one you get by following arch wiki and enabling most "common" perl plugins. Actually I only have one single "plugin" and that is "me"/"mine" plugin. It is a plugin that I wrote myself for myself. It only has everything that I use and nothing else. I admit it might not be prettiest code or it might have negative performance characteristics, although I am not noticing any, but it does only things I want it to do, nothing more and nothing less. 

> Alacritty's performance goals cannot be upheld with arbitrary code that is too poor to land upstream just running around doing random things it shouldn't be doing. If the software is good and the implementation is proper, it can be added to Alacritty, so there's no need for plugins.

While that is one view, especially "plugin-centric" view, there is other important one too. I don't really care about "plugins". When scripting language is added to a program, there are two ways to see that scripting engine: as some "plugin" enabler or as "advanced use and configuration" enabler, eg "user extension" enabler. Emphasis here is on user, which, in this case, is "me", the terminal emulator operating human. 

But you are saying that if there is feature I might want, I should propose it here in issues or just learn rust, implement it to prove my point and try luck with pull request?

Anyway from your sentences I do get impression some level of "scripting" is provided, but not at the level of `urxvt` perl. Are you talking about some additional specifiers in keybindings for example?

> Many of the things handled by URxvt's common plugins are either part of Alacritty already, or currently worked on. Including features like keyboard URL handling, scrollback search and passing the scrollback to a file or buffer.

Indeed, I admit, for many use cases, single default implementation is best approach. And `alacritty` has some interesting builtins. On the other hand, sometimes, there are types of users and there are types of things they want to do, that are not one size fits all friendly at all.

---

### In Conclusion

`alacritty` should never be viewed as `urxvt` analogue. Especially not scripting support.

If one has problem that they would solve by writing `urxvt` script, they are expected to instead to request it as feature here, so it can be added to `alacritty` as core functionalit for benefits of others. Correct?

My most important question is what happens if such feature request gets rejected, it is a dead end then?

**Comment by kchibisov** [comment]
>You declare that is "not your problem" and expect knowledgeable user to restart the terminal (so that said zombie can get "teleported" under init that will reap it without hassle properly) to zap it? Just so that we are clear, yes is valid response here.

Once you spawn a command with a binding in alacritty or by clicking on a url, its PPID will be `1` immediately, so you don't have to restart your terminal at all, I don't get what you mean here. As it was said, it's not our business to supervise them.

>But you are saying that if there is feature I might want, I should propose it here in issues or just learn rust, implement it to prove my point and try luck with pull request?

>Anyway from your sentences I do get impression some level of "scripting" is provided, but not at the level of urxvt perl. Are you talking about some additional specifiers in keybindings for example?

So it's either learning rust to tweak alacritty or learning perl to tweak urxvt, I don't see any difference here? Like you're learning one more language anyway, I don't know perl for example.

> alacritty should never be viewed as urxvt analogue. Especially not scripting support.

Yeah, we don't state that we're urxvt clone.

>If one has problem that they would solve by writing urxvt script, they are expected to instead to request it as feature here, so it can be added to alacritty as core functionalit for benefits of others. Correct?

LIkely, I really don't get what you can ever want from terminal that you require plugin system, and that can't be in core (assuming you can't script some window handling, creating new pty, etc).

>My most important question is what happens if such feature request gets rejected, it is a dead end then?

Use your fork/set of patches? This is how it goes most of the time.

**Comment by etosan** [comment]
> Once you spawn a command with a binding in alacritty or by clicking on a url, its PPID will be 1 immediately, so you don't have to restart your terminal at all, I don't get what you mean here. As it was said, it's not our business to supervise them.

Oh I see, so here is that dreaded doublefork hidden. 

I have yet to use keybinds in `alacritty` itself, it's relatively new feature, I found out only yesterday thanks to this issue.  

Such schemas (keybind to spawn "extension" process/scripts) are usually used at WM level. 

Anyway in 99% of software implementing keybinds (or extensions) that way, the child spawned is the child of the keybinding/extended "host" process (eg. that would be `alacritty` in this case). Almost every WM I know does it this way, so do things like dhcp clients, dns servers or basically anybody else. Are you really sure this is proper way, just because you don't want to "deal" with children ? 

The thing is, it is very easy to doublefork yourself inside the command you are spawning yourself, while it's impossible to "undoublefork" command once it is reparented under init (which makes sense).

It is very weird decision, but if you insist well, I will have to accept it as user - still strange and very "modern", non the less.

I hope at least for daemon mode you are considering, you will have "keep running foreground" option like other proper daemons.

> So it's either learning rust to tweak alacritty or learning perl to tweak urxvt, I don't see any difference here?

Well as a developer you couldn't care less, but as a sysadmin or a power user, knowing bit of perl is much more advantageous to you, than to know rust. Quite lot of managerial scripts and whatnot are done in perl, so at certain level you will meet perl a lot. Of course this is mainly due to history, but such is a reality. Poke around in unix systems long enough and you will eventually hit the "perl line". I am not defending perl, I don't like it very much, but that's simply the way it is.

> Yeah, we don't state that we're urxvt clone.

Nobody said that, but `alacritty`'s primary project goal "very fast rendering, minimal response latency" which is coincidentally same as `urxvt`'s. After all, one of the reasons why `libev` exists is `urxvt` (and amusingly this gave rise to `nodejs` and other stuff branching from there).

With goal like this, it's only given, that you will atract `urxvt` users and expats searching for something more "modern", faster, hw. accelerated or whatever.

From what I've seen `urxvt` users toying with `alacritty` will start to play with idea replacing `urxvt` with it. How well that goes for them depends on how reliant they were on more advanced features.

> Use your fork/set of patches? This is how it goes most of the time.

Understood.

---

Besides final confirmation of "doubleforking commands right away" being set in stone, I guess all my important question are answered. Thank you for your time and explanations guys!

**Comment by chrisduerr** [comment]
> Nobody said that, but alacritty's primary project goal "very fast rendering, minimal response latency" which is coincidentally same as urxvt's.

That's not true at all, unless URxvt is failing horribly at its own goals. It certainly doesn't render very fast.

> The thing is, it is very easy to doublefork yourself inside the command you are spawning yourself, while it's impossible to "undoublefork" command once it is reparented under init (which makes sense).

I mean that's exactly the point. We want nothing to do with our children, so there's no incentive to do anything with them. The farther away from us they are, the better. So there's no reason to care about them being able to undoublefork. In fact, that would be more disadvantage than advantage.
